### DISCUSSION TITLE
Celebrating Charlie: How has NUnit helped you, your career, or your organization?

### DISCUSSION THREAD

**Discussion by SeanKilleen** [nunit/nunit#4283] [General]
We're celebrating the incredible commitment Charlie Poole has made to the NUnit community in his 20+ years of stewarding the NUnit project.

As part of that, we'd love to hear -- how has NUnit helped you, your career, or your organization? We'd love for Charlie to see some of the examples of impact that his work has created.

* Announcement: https://nunit.org/news/update/nunit/2023/01/29/celebrating-charlie-poole.html
* PR for `Assert.Charlie()`: https://github.com/nunit/nunit/pull/4284

**Comment by SeanKilleen** [comment]
I know sometimes people are shy with an empty list, so I'll lead off -- 

While I love xUnit .NET too, NUnit holds a special place in my heart as the framework that allowed me to get started with automated testing, which in turn led to me learning more about TDD and automated testing in general, which has paid so many dividends over the course of my career for myself and those I work with/for.

Using NUnit has allowed me to un-block myself when I'm stuck. It's allowed me to greatly simplify code that's been covered by tests. On one of my favorite projects of my career, I used NUnit heavily, and I received a thank-you letter years later from the current lead developer, who didn't know me, thanking me because the code had run flawlessly for years and it was so easy to confidently make a change they needed to make.

To me, that's what NUnit represents: the increase in my own ability to move with confidence toward the value I want to add. None of that's possible without Charlie steering this ship for so long.

**Comment by mitchdenny** [comment]
@CharliePoole. Thank-you for your contribution to the entire .NET ecosystem. NUnit was the first unit testing framework I used in .NET and was a big factor in me building my skills in the DevOps space to set up continuous integration systems on teams that I worked with. Not only is NUnit a fantastic framework in its own right it really brought unit testing "consciousness" to a lot of people regardless of what frameworks they use today.

Imagine how many "bad commits" have been avoided by the tooling that you created. Your commitment to the project has been inspiring.

Thank-you!

**Comment by pascalberger** [comment]
@CharliePoole Thank you for everything you did for .NET ecosystem. Keeping an OSS project alive and relevant for such a long time is an incredible achievement for you and team you formed around it. NUnit was there from day one of .NET and it's still here, alive and kicking, and changed .NET for the better.

I remember once you asked a question about Cake and starstruck me was happy to answer it and hoped it gave something very small back.

Thank you and all the best for your future endeavors.

**Comment by JasonBock** [comment]
The first testing framework I tried out was JUnit, back in the late 1990s. I didn't use it in my professional career, but the idea stuck with me, though I was still somewhat unconvinced with unit testing in general. When .NET came out, NUnit showed up soon after, and I used NUnit for the first time on a project in 2003. At first, I resisted the entire notion of writing tests, but the team wanted to do it, so I went along for the ride. I remember one day, I changed some code, and when I ran the tests, I found out that some were failing unexpectedly. That's when I became a "convert" to testing. I finally realized having a framework where I could easily run a suite of tests and discover issues before I committed code was a really good thing :).

Over time, Microsoft released MSTest (which was IMO a fiasco), and then later on xUnit started to gain in popularity. However, my choice remains NUnit. I really like the constraint syntax to specify assertions, and the ability to actually provide a message as to **why** a test fails is kind of nice. Furthermore, the `Multiple()` implementation is better than what xUnit offers. I also find xUnit to be rather terse and opinionated in a way that I just don't like. To be fair, I use xUnit when it's the one picked for a project. It's fine. But NUnit is what I gravitate towards.

Thanks @CharliePoole for starting NUnit so many years ago and sticking with it for as long as you did. It's in a great place now, and you should be proud of what you and others on the NUnit team have accomplished.

**Comment by sjrawlins** [comment]
NUnit has played a significant role (off and on again, as my assignments vary) in my career since 2013.  Thank you for this beautiful contribution.
